src/core/compiler.jl

```diff
@@ -134,259 +158,296 @@ end
+    pushfirst!(fdefn_outer.args[2].args, quote
+                 
+                 Turing.eval(:(_compiler_ = deepcopy($compiler)))
+                 fargs = Turing._compiler_[:fargs];
+
+                   # Copy the expr of function definition and callbacks
+                   fdefn_inner = Turing._compiler_[:fdefn_inner];
+                   fdefn_inner_callback1 = Turing._compiler_[:fdefn_inner_callback_1];
+                   fdefn_inner_callback2 = Turing._compiler_[:fdefn_inner_callback_2];
+                   fdefn_inner_callback3 = Turing._compiler_[:fdefn_inner_callback_3];
+
+                    # Add gensym to function name
+                    fname, _, _ = Turing.extractcomponents(fdefn_inner)
+                    fname_gensym = gensym(fname)
+
+                    # Change the name of inner function definition to the one with gensym()
+                    Turing.setfname!(fdefn_inner, fname_gensym)
+                    
+                    Turing.setfname!(fdefn_inner_callback1, fname_gensym)
+                    Turing.setfcall!(fdefn_inner_callback1, fname_gensym)
+                    
+                    Turing.setfname!(fdefn_inner_callback2, fname_gensym)
+                    Turing.setfcall!(fdefn_inner_callback2, fname_gensym)
+                    
+                    Turing.setfname!(fdefn_inner_callback3, fname_gensym)
+                    Turing.setfcall!(fdefn_inner_callback3, fname_gensym)
+
+                    @debug(fdefn_inner_callback1)
+                    @debug(fdefn_inner_callback2)
+                    @debug(fdefn_inner_callback3)
+
+                    @debug fdefn_inner
+               end )
+
+    esc(fdefn_outer)
 end
 
 
 
 ###################
 # Helper function #
 ###################
+function setfcall!_(fexpr::Expr, name::Symbol)
+    if fexpr.head == :function
+        @assert fexpr.args[2].head == :call
+        fexpr.args[2].args[1] = name
+    else
+        @assert length(fexpr.args) > 1
+        setfcall!_(fexpr.args[2], name)
+    end
+end
+function setfcall!(fexpr::Expr, name::Symbol)
+    setfcall!_(fexpr, name)
+end
+function setfname!_(fexpr::Expr, name::Symbol)
+    if fexpr.head == :function
+        fexpr.args[1].args[1] = name
+    else
+        @assert length(fexpr.args) > 1
+        setfname!_(fexpr.args[2], name)
+    end
+end
+function setfname!(fexpr::Expr, name::Symbol)
+    setfname!_(fexpr, name)
+end
+
+"""
+  constructfunc(fname::Symbol, fargs, fbody)
+
+Construct a function.
+"""
+function constructfunc(fname::Symbol, fargs, fbody)
+    fdefn = Expr(:function, 
+                 Expr(:call, 
+                      fname,
+                      fargs...
+                     ),
+                 deepcopy(fbody)
+                )
+    return fdefn
+end
+
+"""
+  insertvarinfo(fexpr::Expr)
+
+Insert `_lp=0` to function call and set `vi.logp=_lp` inplace at the end.
+"""
+insertvarinfo(fexpr::Expr) = insertvarinfo!(deepcopy(fexpr))
+function insertvarinfo!(fexpr::Expr)
```

Comment:
We decided to not change the return statement anymore. So that the model definition specified by the user is as close to the model function as possible. VarInfo is now only changed in-place and not returned.

Reply:
Ah okay, makes sense. Is there not a chance of subtle bugs if we don't place a `vi.logp = _lp` statement before a `return` that appears mid-function definition?
ie. as I understand it, something like
```julia
@model foo(x, y) = begin
    x ~ Normal(0, 1)
    if x > 0
        return x
    end
    y ~ Normal(x, 1)
    if y > x
        return y
    end
end
```
won't get a `vi.logp = _lp` statement until the end of the model, so the `logp` field of `VarInfo` will never be set?

Reply:
Good point, this is a problem of the current compiler too.

Reply:
this will indeed cause issues

Reply:
Okay. Even if we don't resolve this issue in this PR, we please either
- resolve this in this PR and add a some tests, or
- add a test that we know fails using the `@test_skip` macro so that we don't forget to come back to this?

Reply:
I think we should resolve the issue in this PR and add tests.